File: crates/livraria-domain/tests/conformance.rs
//! Lado NATIVO do teste de conformidade (ADR-0022). Roda os mesmos vetores que o
//! `@livraria/domain` (WASM) roda em `conformance.mjs`: mesma entrada → mesma saída.
//! Se ambos passam sobre os mesmos vetores, PDV (nativo) e Escritório (WASM) são idênticos.

use livraria_domain::{dinheiro::Dinheiro, estoque};
use serde_json::Value;

const VECTORS: &str =
    include_str!("../../../specs/008-escritorio-espelho-pdv/contracts/conformance-vectors.json");

fn casos<'a>(v: &'a Value, chave: &str) -> &'a Vec<Value> {
    v[chave].as_array().unwrap()
}

fn i(v: &Value) -> i64 {
    v.as_i64().unwrap()
}

#[test]
fn conformidade_nativa() {
    let v: Value = serde_json::from_str(VECTORS).unwrap();

    for c in casos(&v, "clamp_baixa_venda") {
        assert_eq!(estoque::clamp_baixa_venda(i(&c["in"][0]), i(&c["in"][1])), i(&c["out"]));
    }
    for c in casos(&v, "baseline_saldo_inicial") {
        assert_eq!(estoque::baseline_saldo_inicial(i(&c["in"][0]), i(&c["in"][1])), i(&c["out"]));
    }
    for c in casos(&v, "diferenca_contagem") {
        assert_eq!(estoque::diferenca_contagem(i(&c["in"][0]), i(&c["in"][1])), i(&c["out"]));
    }
    for c in casos(&v, "custo_medio_apos_entrada") {
        let got = estoque::custo_medio_apos_entrada(
            i(&c["in"][0]),
            Dinheiro::de_centavos(i(&c["in"][1])),
            i(&c["in"][2]),
            Dinheiro::de_centavos(i(&c["in"][3])),
        )
        .centavos();
        assert_eq!(got, i(&c["out"]));
    }
    for c in casos(&v, "parse_brl") {
        let got = Dinheiro::parse_brl(c["in"].as_str().unwrap()).unwrap().centavos();
        assert_eq!(got, i(&c["out"]));
    }
    for c in casos(&v, "to_brl") {
        assert_eq!(Dinheiro::de_centavos(i(&c["in"])).to_brl(), c["out"].as_str().unwrap());
    }
    for c in casos(&v, "recompor_ledger") {
        let movs: Vec<(i64, Option<i64>)> = c["in"]
            .as_array()
            .unwrap()
            .iter()
            .map(|m| (i(&m[0]), if m[1].is_null() { None } else { Some(i(&m[1])) }))
            .collect();
        let (saldo, medio) = estoque::recompor_ledger(&movs);
        assert_eq!(saldo, i(&c["saldo"]));
        assert_eq!(medio.centavos(), i(&c["custo"]));
    }
}

File: specs/008-escritorio-espelho-pdv/contracts/conformance.mjs
// Lado WASM do teste de conformidade (ADR-0019). Carrega @livraria/domain
// (packages/domain, gerado pelo CI) e roda os MESMOS vetores que o teste nativo
// (crates/livraria-domain/tests/conformance.rs). Node: `node conformance.mjs`.
import { readFileSync } from "node:fs";
import init, * as dom from "../../../packages/domain/index.js";

const bytes = readFileSync(new URL("../../../packages/domain/index_bg.wasm", import.meta.url));
await init({ module_or_path: bytes });

const vetores = JSON.parse(readFileSync(new URL("./conformance-vectors.json", import.meta.url)));

let falhas = 0;
const eq = (nome, obtido, esperado) => {
  const o = JSON.stringify(obtido);
  const e = JSON.stringify(esperado);
  if (o !== e) {
    console.error(`  ✗ ${nome}: obtido ${o}, esperado ${e}`);
    falhas++;
  }
};

for (const c of vetores.clamp_baixa_venda) eq("clamp_baixa_venda", dom.clamp_baixa_venda(...c.in), c.out);
for (const c of vetores.baseline_saldo_inicial) eq("baseline_saldo_inicial", dom.baseline_saldo_inicial(...c.in), c.out);
for (const c of vetores.diferenca_contagem) eq("diferenca_contagem", dom.diferenca_contagem(...c.in), c.out);
for (const c of vetores.custo_medio_apos_entrada) eq("custo_medio_apos_entrada", dom.custo_medio_apos_entrada(...c.in), c.out);
for (const c of vetores.parse_brl) eq("parse_brl", dom.parse_brl(c.in), c.out);
for (const c of vetores.to_brl) eq("to_brl", dom.to_brl(c.in), c.out);
for (const c of vetores.recompor_ledger) {
  const r = dom.recompor_ledger(c.in);
  eq("recompor_ledger.saldo", r.saldo, c.saldo);
  eq("recompor_ledger.custo", r.custo_medio_centavos, c.custo);
}
for (const c of vetores.turno_proximo_numero) eq("turno_proximo_numero", dom.turno_proximo_numero(c.in), c.out);
for (const c of vetores.turno_encerrar) eq("turno_encerrar", dom.turno_encerrar(...c.in).diferencaCentavos, c.out);
for (const c of vetores.troco_venda) {
  const itens = c.itens.map(([precoCentavos, qtd]) => ({ precoCentavos, qtd }));
  const pagamentos = c.pagamentos.map(([formaId, valorCentavos]) => ({ formaId, valorCentavos }));
  eq("troco_venda", dom.troco_venda(itens, pagamentos), c.out);
}
for (const c of vetores.contagem_efetiva) eq("contagem_efetiva", dom.contagem_efetiva(c.in[0], c.in[1], c.in[2]), c.out);
for (const c of vetores.resumir_inventario) eq("resumir_inventario", dom.resumir(c.in), c.out);

if (falhas > 0) {
  console.error(`\n${falhas} divergência(s) PDV↔WASM.`);
  process.exit(1);
}
console.log("✓ conformidade PDV↔WASM: todos os vetores conferem (nativo == WASM).");
